File: python/slack.py
"""Slack messaging via API."""
from typing import Dict, Any, List, Optional
import aiohttp


class SlackSkill:
    def __init__(self, token: str):
        self.token = token
        self.base_url = "https://slack.com/api"

    async def send_message(self, channel: str, text: str,
                           blocks: Optional[List[Dict]] = None,
                           thread_ts: Optional[str] = None) -> Dict[str, Any]:
        headers = {"Authorization": f"Bearer {self.token}"}
        payload = {"channel": channel, "text": text, "thread_ts": thread_ts}
        if blocks:
            payload["blocks"] = blocks
        async with aiohttp.ClientSession() as session:
            async with session.post(f"{self.base_url}/chat.postMessage", json=payload, headers=headers) as resp:
                result = await resp.json()
                if result.get("ok"):
                    return {"success": True, "message": result}
                return {"success": False, "error": result.get("error", "Unknown error")}

File: power-agent/skills/slack.js
// Slack Skill — Send messages via Slack API
export class SlackSkill {
  name = "slack";
  description = "Send messages to Slack channels";

  tools() {
    return [
      {
        name: "send_message",
        description: "Send a message to a Slack channel",
        inputSchema: {
          type: "object",
          properties: {
            channel: { type: "string", description: "Slack channel ID or name (e.g. C12345 or #general)" },
            text: { type: "string", description: "Message text to send" },
            token: { type: "string", description: "Slack Bot User OAuth Token (xoxb-...)" }
          },
          required: ["channel", "text", "token"]
        }
      }
    ];
  }

  async send_message(args) {
    try {
      const { channel, text, token } = args;
      const response = await fetch("https://slack.com/api/chat.postMessage", {
        method: "POST",
        headers: {
          "Authorization": `Bearer ${token}`,
          "Content-Type": "application/json"
        },
        body: JSON.stringify({ channel, text })
      });
      const result = await response.json();
      if (result.ok) {
        return { success: true, data: { ts: result.ts, channel: result.channel } };
      }
      return { success: false, error: result.error || "Unknown Slack API error" };
    } catch (e) {
      return { success: false, error: e.message };
    }
  }
}
